### ISSUE TITLE
Automatically truncate long tweets

### ISSUE THREAD

**Issue by vestingz** [IBBoard/cawbird#398]
At the moment, users have to manually truncate their tweets if they exceed the character limit, which is slightly annoying.

A possible solution would be to automatically truncate the text at the char limit and add a suffix indicating a multi-part tweet (e.g. 1/2 and 2/2). Regarding the specific cutting point, it could be the last available white space before meeting the char limit, avoiding cutting apart words.

**Comment by IBBoard** [comment]
I agree that we could do more for multi-part tweets, but I'm not sure that splitting it for the user is the right way. Amongst other problems, I can see:

* Sometimes you're better moving an entire sentence or paragraph to a new post, so making the assumption that it's just the trailing words at whitespace splits may not hold a lot of the time
* Sometimes you want to add more to the thread later (e.g. a follow-up tweet with images) so the numbering could be wrong 
* The widgets for the split are going to be awkward - how do we show that there's two parts? Or three? And how does that dynamically shift when the user edits and rewords earlier in the tweet?

**Comment by vestingz** [comment]
Thanks for picking up the idea :)

>     * Sometimes you're better moving an entire sentence or paragraph to a new post, so making the assumption that it's just the trailing words at whitespace splits may not hold a lot of the time
True. But maybe this could just be solved by letting the user choose their delimiter, with one possibility being a non-whitespace string with an dot in the end. But I am sure there are plenty of edge cases, you are absolutely right.

>     * Sometimes you want to add more to the thread later (e.g. a follow-up tweet with images) so the numbering could be wrong
Also true. But could also be "solved" by a respective option to either quote the total number of tweets or not.
 
>     * The widgets for the split are going to be awkward - how do we show that there's two parts? Or three? And how does that dynamically shift when the user edits and rewords earlier in the tweet?
You already have a dynamic character counter in the tweet compose window. Couldn't you just show the number of split-parts next to that number, provided the total character count exceeds the maximum tweet length, like so:  `677/3`? That would indicate the total number of characters and the amount of split parts.
If you are referring to the text in the compose window itself, there could be a different background color each 280 characters. But I think that something like this would also not be 100% necessary. 
In any case, there is a limit to work around the platforms inherent limitations. Automatically truncating tweets can hardly be much more than a dirty hack.

**Comment by CodedOre** [comment]
> The widgets for the split are going to be awkward - how do we show that there's two parts? Or three? And how does that dynamically shift when the user edits and rewords earlier in the tweet?

I recently made an mockup for a "Post Creation" dialog for the redesign which takes threads into account:

![create-thread-mockup](https://user-images.githubusercontent.com/47981497/132694166-6eab41e5-098a-4ea5-82aa-5cbba2e4586d.png)

I think the option to create multiple posts at once should be something that should be included in the Redesign, since Twitter also [provides this functionality.](https://user-images.githubusercontent.com/47981497/132694436-64ac316f-c102-4512-8ded-0a695f7ed866.png)

But doing it automatically is not necessary IMHO.

**Comment by vestingz** [comment]
> I recently made an mockup for a "Post Creation" dialog for the
> redesign which takes threads into account:
> 
> ![create-thread-mockup](https://user-images.githubusercontent.com/47981497/132694166-6eab41e5-098a-4ea5-82aa-5cbba2e4586d.png)
>

Looks really nice :)

>  I think the option to create multiple posts at once should be
> something that should be included in the Redesign, since Twitter also
> [provides this
> functionality.](https://user-images.githubusercontent.com/47981497/132694436-64ac316f-c102-4512-8ded-0a695f7ed866.png)
>
>  But doing it automatically is not necessary IMHO.

Yeah, doesn't need to be automatic. Just at the moment making a thread 
involves some amount of copy-pasta and posting some 600 chars is 
generally fiddly.

**Comment by Nardol** [comment]
Wy not using a service like [Twishort](https://twishort.com/dev)?
It would allow users to type their tweet and if exceeding the limit, a link could be created.

**Comment by CodedOre** [comment]
> Wy not using a service like Twishort?
> It would allow users to type their tweet and if exceeding the limit, a link could be created.

Using an external service like Twishort would introduce an unnecessary dependency and would likely also not be the behaviour most of our user would like to see.
I think adding the option to compose threads in Cawbird 2.0 is the best solution here.

**Comment by IBBoard** [comment]
Also, services like Twishort die and get closed down. I've seen various posts from people that are no longer fully available because either they used a "long tweet" service or they used one of the "post a tweet and an image, but not on Twitter" services.

It's 50% of the same reason why we won't support automated link shortening.

Also also, people have to click a link to open it. The whole point of Cawbird is that people can *avoid* using their browser for viewing Twitter!
